File: p2pnet/android/app/src/main/java/com/example/p2pnet/util/Crypto.kt
package com.example.p2pnet.util

import java.security.MessageDigest
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * 加密工具类
 * p2pnet 认证流程：
 * 1. pw_hash = SHA256(password + salt)  (password+salt as UTF-8 bytes, output as hex)
 * 2. response = HMAC-SHA256(hex2bin(pw_hash), hex2bin(challenge))
 */
object Crypto {

    private const val HMAC_SHA256 = "HmacSHA256"
    private const val SHA256 = "SHA-256"

    fun sha256(input: String): String {
        return sha256Bytes(input).toHexString()
    }

    fun sha256Bytes(input: String): ByteArray {
        return MessageDigest.getInstance(SHA256).digest(input.toByteArray(Charsets.UTF_8))
    }

    fun hmacSha256(keyBytes: ByteArray, dataBytes: ByteArray): ByteArray {
        val mac = Mac.getInstance(HMAC_SHA256)
        val secretKey = SecretKeySpec(keyBytes, HMAC_SHA256)
        mac.init(secretKey)
        return mac.doFinal(dataBytes)
    }

    fun hmacSha256Hex(key: ByteArray, data: ByteArray): String {
        return hmacSha256(key, data).toHexString()
    }

    fun computeAuthResponse(password: String, salt: String, challenge: String): String {
        val pwHash = sha256(password + salt)
        val pwHashBytes = hexToBytes(pwHash)
        val challengeBytes = hexToBytes(challenge)
        val hmacBytes = hmacSha256(pwHashBytes, challengeBytes)
        return hmacBytes.toHexString()
    }

    private fun ByteArray.toHexString(): String = joinToString("") { "%02x".format(it) }

    fun hexToBytes(hex: String): ByteArray {
        require(hex.length % 2 == 0) { "Hex string must have even length" }
        return ByteArray(hex.length / 2) { i ->
            hex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        }
    }
}

File: p2pnet/ios/p2pnet/Util/Crypto.swift
import Foundation
import CommonCrypto

enum Crypto {
    static func sha256(_ input: String) -> String {
        let data = Data(input.utf8)
        var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        data.withUnsafeBytes { ptr in
            _ = CC_SHA256(ptr.baseAddress, CC_LONG(data.count), &hash)
        }
        return hash.map { String(format: "%02x", $0) }.joined()
    }

    static func sha256Bytes(_ input: String) -> Data {
        let data = Data(input.utf8)
        var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        data.withUnsafeBytes { ptr in
            _ = CC_SHA256(ptr.baseAddress, CC_LONG(data.count), &hash)
        }
        return Data(hash)
    }

    static func hmacSHA256(key: Data, data: Data) -> Data {
        var hmac = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        key.withUnsafeBytes { keyPtr in
            data.withUnsafeBytes { dataPtr in
                CCHmac(CCHmacAlgorithm(kCCHmacAlgSHA256), keyPtr.baseAddress, key.count, dataPtr.baseAddress, data.count, &hmac)
            }
        }
        return Data(hmac)
    }

    static func hmacSHA256Hex(key: Data, data: Data) -> String {
        return hmacSHA256(key: key, data: data).map { String(format: "%02x", $0) }.joined()
    }

    static func hexToBytes(_ hex: String) -> Data {
        var bytes = [UInt8]()
        for i in stride(from: 0, to: hex.count, by: 2) {
            let start = hex.index(hex.startIndex, offsetBy: i)
            let end = hex.index(start, offsetBy: 2)
            let byte = UInt8(hex[start..<end], radix: 16)!
            bytes.append(byte)
        }
        return Data(bytes)
    }

    static func computeAuthResponse(password: String, salt: String, challenge: String) -> String {
        let pwHash = sha256(password + salt)
        let pwHashBytes = hexToBytes(pwHash)
        let challengeBytes = hexToBytes(challenge)
        let hmac = hmacSHA256(key: pwHashBytes, data: challengeBytes)
        return hmac.map { String(format: "%02x", $0) }.joined()
    }

    static func generateRandomBase64(_ byteLen: Int) -> String {
        var bytes = [UInt8](repeating: 0, count: byteLen)
        _ = SecRandomCopyBytes(kSecRandomDefault, byteLen, &bytes)
        return Data(bytes).base64EncodedString()
    }
}